asyncstdlib/itertools.py

```diff
@@ -141,35 +141,38 @@ async def accumulate(iterable, function, *, initial):
             yield value
 
 
-async def chain(*iterables: AnyIterable[T]) -> AsyncIterator[T]:
+class chain(Generic[T], AsyncIterator[T]):
     """
     An :term:`asynchronous iterator` flattening values from all ``iterables``
 
     The resulting iterator consecutively iterates over and yields all values from
     each of the ``iterables``. This is similar to converting all ``iterables`` to
     sequences and concatenating them, but lazily exhausts each iterable.
     """
-    for iterable in iterables:
-        async with ScopedIter(iterable) as iterator:
-            async for item in iterator:
-                yield item
-
-
-@public_module(__name__, "chain.from_iterable")
-async def chain_from_iterable(
-    iterable: AnyIterable[AnyIterable[T]],
-) -> AsyncIterator[T]:
-    """
-    Alternate constructor for :py:func:`~.chain` that lazily exhausts iterables as well
-    """
-    async with ScopedIter(iterable) as iterables:
-        async for sub_iterable in iterables:
-            async with ScopedIter(sub_iterable) as iterator:
-                async for item in iterator:
-                    yield item
-
 
-chain.from_iterable = chain_from_iterable  # type: ignore
+    __slots__ = "_impl",
+
+    def __init__(self, *iterables: AnyIterable[T]):
+        async def impl():
+            for iterable in iterables:
+                async with ScopedIter(iterable) as iterator:
+                    async for item in iterator:
+                        yield item
+        self._impl = impl()
+
+    @classmethod
+    async def from_iterable(cls: type[Any], iterable: AnyIterable[AnyIterable[T]]) -> AsyncIterator[T]:
```

Comment:
At least MyPy is happy if you don't annotate `cls` at all. Since the class isn't actually used, a `staticmethod` should work as well:
```suggestion
    @staticmethod
    async def from_iterable(iterable: AnyIterable[AnyIterable[T]]) -> AsyncIterator[T]:
```